test/framework/src/main/java/org/elasticsearch/index/shard/IndexShardTestCase.java

```diff
@@ -0,0 +1,375 @@
+        Store store = createStore(indexSettings, shardPath);
+        IndexCache indexCache = new IndexCache(indexSettings, new DisabledQueryCache(indexSettings), null);
+        MapperService mapperService = MapperTestUtils.newMapperService(createTempDir(), indexSettings.getSettings());
+        for (ObjectObjectCursor<String, MappingMetaData> typeMapping : indexMetaData.getMappings()) {
+            mapperService.merge(typeMapping.key, typeMapping.value.source(), MapperService.MergeReason.MAPPING_RECOVERY, true);
+        }
+        SimilarityService similarityService = new SimilarityService(indexSettings, Collections.emptyMap());
+        final IndexEventListener indexEventListener = new IndexEventListener() {
+        };
+        final Engine.Warmer warmer = searcher -> {
+        };
+        IndicesFieldDataCache indicesFieldDataCache = new IndicesFieldDataCache(nodeSettings, new IndexFieldDataCache.Listener() {
+        });
+        IndexFieldDataService indexFieldDataService = new IndexFieldDataService(indexSettings, indicesFieldDataCache,
+            new NoneCircuitBreakerService(), mapperService);
+
+        return new IndexShard(routing, indexSettings, shardPath, store, indexCache, mapperService, similarityService, indexFieldDataService,
+            null, indexEventListener, indexSearcherWrapper, threadPool, BigArrays.NON_RECYCLING_INSTANCE, warmer, Collections.emptyList(),
+            Arrays.asList(listeners));
+    }
+
+    protected IndexShard reinitShard(IndexShard current, IndexingOperationListener... listeners) throws IOException {
+        final ShardRouting shardRouting = current.routingEntry();
+        return reinitShard(current, ShardRoutingHelper.initWithSameId(shardRouting,
+            shardRouting.primary() ? RecoverySource.StoreRecoverySource.EXISTING_STORE_INSTANCE : RecoverySource.PeerRecoverySource.INSTANCE
+            ), listeners);
+    }
+
+    protected IndexShard reinitShard(IndexShard current, ShardRouting routing, IndexingOperationListener... listeners) throws IOException {
+        closeShards(current);
+        return newShard(routing, current.shardPath(), current.indexSettings().getIndexMetaData(), null, listeners);
+    }
+
+    protected IndexShard newStartedShard() throws IOException {
+        return newStartedShard(randomBoolean());
+    }
+
+    protected IndexShard newStartedShard(boolean primary) throws IOException {
+        IndexShard shard = newShard(primary);
+        if (primary) {
+            recoveryShardFromStore(shard);
+        } else {
+            recoveryEmptyReplica(shard);
+        }
+        return shard;
+    }
+
+    protected void closeShards(IndexShard... shards) throws IOException {
+        closeShards(Arrays.asList(shards));
+    }
+
+    protected void closeShards(Iterable<IndexShard> shards) throws IOException {
+        for (IndexShard shard : shards) {
+            if (shard != null) {
+                shard.close("test", false);
```

Comment:
try finally  here please since if close fails we don't release lock etc which can be missleading